0010-B-object-storage-building-block.md

```diff
@@ -0,0 +1,264 @@
+
+When invoking `GetObject` or `SetObject`, the output and input streams (respectively) are created outside of the state store and are owned by the caller. The `GetObject` and `SetObject` methods synchronously return only after all data has been written or read (respectively), and after that the caller can call `Close()` on the streams (if necessary/appropriate).
+
+Notes:
+
+- This building block will be stream-first. It will be working with input and output data as binary streams, and will not perform any transformation or encoding on the data. Thanks to using streams, there are no limits on the size of the input/output data, bypassing `MaxBodySize`.
+- Dapr only supports the "last-write-wins" concurrency pattern for object storage (unlike with state stores). This makes sense considered the intended usage of object storage, which is storing documents that usually do not change. When documents do change, using last-write-wins is consistent with how regular filesystems work, for example.
+- Dapr will not calculate and store the checksum of objects. This is because checksums must be transmitted as header, but Dapr stores data in the state store in a streaming way, so it can't compute the full checksum until the end of the stream. Apps that want to store checksums (such as the `Content-MD5` header) should compute it beforehand and submit it as metadata header.
+
+### HTTP APIs
+
+Applications that interact with object storage using HTTP can leverage these APIs, that follow a more REST-like pattern than the typical state store APIs.
+
+> These APIs have been modeled after the APIs supported by AWS S3 and Azure Blob Storage.
+
+#### Retrieving an object
+
+To retrieve an object, clients make a GET request:
+
+```text
+GET http://localhost:<port>/v1.0-alpha1/objects/<component>/<key>
+```
+
+Where `<component>` is the name of the object store component, and `<key>` is the key (name) of the object.
+
+A successful response will be:
+
+- Status: `200 OK`
+- Headers: each metadata key is a header; see below for a comment on headers. Additionally, the `ETag` is passed as response header too.
+- Body: the body of the response is the raw bytes of the object, which are sent to the client in a streamed way, as soon as they are retrieved from the state store (that is: Dapr does not buffer the data in-memory before sending it to the client).
+
+An error response will have:
+
+- Status:
+  - `404 Not Found` for objects not found
+  - `400 Bad Request` if the state store doesn't exist or the key parameter is missing
+  - `500 Internal Server Error` for all other errors
+- Headers: response will have `Content-Type: application/json`
+- Body: a JSON-encoded payload including more details on the error
+
+#### Storing an object
+
+To store an object, clients make a PUT request:
```

Comment:
If a user has a state store component that is used both for KV and Blob, will a key from one API override the other? If so, this might create inconsistencies that are very hard to reason about, especially during runtime.

Reply:
We should recommend users to not use the same storage accounts / buckets with multiple building blocks.

I don't see this much different than someone using for example Postgres as a Dapr state store and also interacting with the database directly using the Postgres APIs.

Reply:
I see a great difference, because in a Dapr state Postgres + Direct Postgres scenario saving the same key would actually work as expected by default whereas with Dapr state S3 and Dapr object S3 you get values being overridden by default.

Reply:
I am not following, sorry.

What do you mean with:

> with Dapr state S3 and Dapr object S3 you get values being overridden by default.

Reply:
If you're calling a Dapr S3 state store for saving state with key `1`, then call a Dapr object store to store an object with key `1`, a GET operation for the S3 state store would return the data saved by the object store.

Reply:
Right, that's the same as using Dapr Postgres state store to save state with key `1`, then using the Postgres client to query for key `1`. It is working as intended IMHO. (although keep in mind that the Dapr state stores have a [prefix](https://docs.dapr.io/developing-applications/building-blocks/state-management/howto-share-state/#specifying-a-state-prefix-strategy) by default, so users who do want to overwrite `1` would actually need to write `prefix/1`)

The documentation should not encourage users to use multiple building blocks to access the same resources.

Reply:
Save state with Dapr for key `1` results in a key in the format of `app-id||1`. Saving state with an SDK for key `1` would result in a key of `1`.

So if a user saves two different values, with and without Dapr, and performs a GET call with and without Dapr, the values would be taken from two different keys, as expected. That's not the case with Dapr state and object storage, where both operations target `app-id||1` which is the same key.

Reply:
> The documentation should not encourage users to use multiple building blocks to access the same resources.

Best to incorporate into the proposal as a note then.